Example: Add datasource-streaming-backend-websocket

oshirohugo opened the pull request:
Add a datasource plugin example capable of opening a stream connection to the backend and connect to a websocket service
Closes #87

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana-plugin-examples?pullRequest=179) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana-plugin-examples?pullRequest=179) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana-plugin-examples?pullRequest=179) it.</sub>

andresmgot commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/src/types.ts:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Value that is used in the backend, but never sent over HTTP to the frontend
+ */
+export interface MySecureJsonData {
```
this interface is not used

andresmgot reviewed (commented):
Great progress here! So far I have a few small suggestions.

I am no expert for the streaming feature but I tested the plugin and worked fine :+1:

oshirohugo commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/pkg/plugin/datasource.go:
```diff
@@ -0,0 +1,191 @@
+			var msg Message
+			json.Unmarshal([]byte(rawMsg), &msg)
+
+			value := msg.Value*(d.upperLimit-d.lowerLimit) + d.lowerLimit
```
`msg.Value` comes from the websocket connection. It's a random value, and the `upperLimit` and `loweLimit` are limit the range of the random value. My goal here was to exemplify how to pass data from the frontend to backend when a connection is made.

andresmgot replied:
can you add a comment with this?

oshirohugo commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/pkg/plugin/datasource.go:
```diff
@@ -0,0 +1,191 @@
+		return err
+	}
+
+	// simple validation of provided parameters
+	if q.UpperLimit > q.LowerLimit {
+		d.upperLimit = q.UpperLimit
+		d.lowerLimit = q.LowerLimit
+	}
```
What would be the right way to pass a value from frontend to backend in a stream plugin after the plugin creation?

andresmgot replied:
is it not part of the `RunStream` request? there is a `req *backend.RunStreamRequest`. This should be encoded somehow there.

Note that with this, two different queries for the same data source will edit the same data source properties (only the values of the last one executed will be saved).

oshirohugo replied:
I changed this, so each query will generate a different stream channel path and a different websocket connection. I'm taking the parameters from the channel path, like it's done in the MQTT plugin

xnyo reviewed (commented):
Awesome work! :rocket: :rocket:   I tested it and it's working great.

I have left some comments about some typos and some missing error handling.

There's one issue with a goroutine that's not closed and I am not too sure how to solve it properly, just flagging it for now. I can take a better look next week :smiley:

oshirohugo commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/pkg/websocket/client.go:
```diff
@@ -0,0 +1,65 @@
+		}
+	}()
+	return messages
+}
```
Nice catch! I solve this closing the connection from the caller when `ctx.Done()`. This will case an error on `websocket.Message.Receive(c.conn, &message)`, because the connection is close, then the loop will continue to the next case which is the `ctx.Done()`.
Here I changed a little bit your suggestion and `return` instead of `break` to avoid the `outer` label

oshirohugo replied:
After some changes in the code, I've introduced one connection per query (like it was done in mqtt plugin) to fulfill Andres concerns about changing the datasource. 
Doing that I was having situation where the connection was close, to I would receive an err in `websocket.Message.Receive` but the `ctx.Done()` fired, so I've introduced a new channel `stopChan` that would be closed when the connection is closed and that would stop the loop

oshirohugo commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/src/datasource.ts:
```diff
@@ -0,0 +1,14 @@
+    super(instanceSettings);
+  }
+
+  getDefaultQuery(app: CoreApp): Partial<MyQuery> {
+    return DEFAULT_QUERY;
+  }
```
This is an easy way to have default values for my query, it's from `DataSourceApi`

andresmgot commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/pkg/plugin/datasource.go:
```diff
@@ -172,6 +167,32 @@ func (d *Datasource) PublishStream(context.Context, *backend.PublishStreamReques
+		return fmt.Errorf("invalid path: %s", req.Path)
+	}
+
+	q := Query{
+		LowerLimit: chunks[1],
+		UpperLimit: chunks[2],
+	}
```
this may work but I see that there ris a req.Data that seems like a better place to put query inputs. See for example how Loki is using it:

https://github.com/grafana/grafana/blob/main/pkg/tsdb/loki/streaming.go#L68

oshirohugo replied:
To be able to set `req.Data` I had to add frontend code. I've added a new implementation to `query` to create the stream. That made the `Query` implementation in the backend unnecessary to create the stream, so I removed it, which made the implementation of `backend.QueryDataHandler` unnecessary as well.

andresmgot commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/src/components/QueryEditor.tsx:
```diff
@@ -1,24 +1,33 @@
     <div className="gf-form">
       <InlineField label="Lower Limit" labelWidth={16} tooltip="Random numbers lower limit">
-        <Input onChange={handleEvent('lowerLimit', parseFloat)} onBlur={onRunQuery} value={lowerLimit || ''} />
+        <Input onChange={onLowerLimitChange} onBlur={onRunQuery} value={lowerLimit || ''} />
```
this should be `type="number"` btw

andresmgot reviewed (commented):
I tried this again and it seems that the inputs are being ignored? It always shows points between 0 and 1:

![Screenshot from 2023-07-21 10-21-30](https://github.com/grafana/grafana-plugin-examples/assets/4025665/a84c38ed-d599-41f0-857e-17f66a94d5f1)

oshirohugo commented:
@andresmgot @xnyo this is ready for review again
Current changes:
- Added `query` implementation in `datasource.ts` (FE)
- Removed `Query` implementation in `datasource.go` (BE)
- Parsing query parameters from `req.Data` in `RunStream`
- Changed all query parameters both BE and FE to number instead of string

andresmgot commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/src/datasource.ts:
```diff
@@ -0,0 +1,42 @@
+import { MyQuery, MyDataSourceOptions, DEFAULT_QUERY } from './types';
+
+export class DataSource extends DataSourceWithBackend<MyQuery, MyDataSourceOptions> {
+  constructor(instanceSettings: DataSourceInstanceSettings<MyDataSourceOptions>) {
+    super(instanceSettings);
+  }
```
nit: if you are not modifying anything you can skip this?

oshirohugo replied:
I've notice this code is comes from `npx @grafana/create-plugin@latest`.
Should we also remove it from the template later?

oshirohugo replied:
Actually this is necessary indeed. I've removed it and the datasource frontend broke

andresmgot reviewed (commented):
Thanks for the changes! Code wise I only have some minor suggestions left.

In any case, it seems that the limits are still not working (at least for me). No matter what I introduce in the lower or upper limit, it still return values between 0-1. Am I missing something?

oshirohugo commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/src/datasource.ts:
```diff
@@ -0,0 +1,42 @@
+        addr: {
+          scope: LiveChannelScope.DataSource,
+          namespace: this.uid,
+          path: `my-ws/custom-${query.lowerLimit}-${query.upperLimit}`, // this will allow each new query to create a new connection
```
The path doesn't need to include the limit, but the path needs to change every time the query changes. A new query should create a new connection and the way of doing it is by creating a new path. A prettier way of doing the same thing is by creating a hash from the query parameters (that is what loki does, e.g.: 
https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/streaming.ts#L68

https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/streaming.ts#L21-L28)

I did like this because it seems simpler

oshirohugo commented:
> Thanks for the changes! Code wise I only have some minor suggestions left.
> 
> In any case, it seems that the limits are still not working (at least for me). No matter what I introduce in the lower or upper limit, it still return values between 0-1. Am I missing something?
![Screenshot from 2023-07-24 14-25-35](https://github.com/grafana/grafana-plugin-examples/assets/5558280/b0e4fae4-599f-4548-9ade-756c9ccee46a)

It seems to be working, but I'm trying to reproduce your error

andresmgot reviewed (approved):
I rebuilt the images and now it's working as expected :+1: LGTM! Let's see if @xnyo finds something else.

xnyo commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/pkg/plugin/datasource.go:
```diff
@@ -154,38 +93,65 @@ func (d *Datasource) CheckHealth(_ context.Context, req *backend.CheckHealthRequ
 func (d *Datasource) RunStream(ctx context.Context, req *backend.RunStreamRequest, sender *backend.StreamSender) error {
-	messages := d.wsClient.Read()
+	q := Query{}
+	json.Unmarshal(req.Data, &q)
```
Missing error handling:

```suggestion
	if err := json.Unmarshal(req.Data, &q); err != nil {
		return fmt.Errorf("json unmarshal: %w", err)
	}
```

xnyo commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/pkg/plugin/datasource.go:
```diff
@@ -154,38 +93,65 @@ func (d *Datasource) CheckHealth(_ context.Context, req *backend.CheckHealthRequ
 			var msg Message
-			json.Unmarshal([]byte(rawMsg), &msg)
+			if err := json.Unmarshal([]byte(rawMsg), &msg); err != nil {
+				Logger.Error("Failed to unmarshal message", "error", err)
```
Here it should skip execution of the body, so you should change this to:

```suggestion
				Logger.Error("Failed to unmarshal message", "error", err)
				break	
```

oshirohugo commented:
@andresmgot you were right, there was indeed a problem with the range. I could reproduce the error after changing the query multiple times. The connection was reused, which was causing many problems. Sometimes the connection would die out of nowhere and a re-connection would happen with default data, causing the issue you've mentioned. 
I changed the code, so every `RunStream` would create a new connection and the connection would be closed when `RunStream` returns

I've also simplified the error treatment. Since this is an example, for any error, I'm terminating the streaming. This should address the problem that @xnyo mentioned. But let me know if you prefer to add logs.

I've simplified the websocket client code and usage, by using `"github.com/gorilla/websocket"` instead of  `"golang.org/x/net/websocket"`. I was having many problem with it during my investigation  and according to its docs:
> This package currently lacks some features found in an alternative and more actively maintained WebSocket package

And gorilla seems to be back again!
https://github.com/gorilla/mux/issues/707#issuecomment-1629838706

andresmgot reviewed (approved):
:+1: cool, sounds good to me

xnyo commented on examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/pkg/websocket/client.go:
```diff
@@ -0,0 +1,44 @@
+	"github.com/grafana/grafana-plugin-sdk-go/backend/log"
+)
+
+var Logger = log.DefaultLogger
```
Nit: This variable is now unused

xnyo reviewed (approved):
Awesome work! LGTM!
